docs/async/async-csharp-vb.rst

```diff
@@ -0,0 +1,199 @@
+By `Phillip Carter`_
+
+Async programming in C# and VB share a language-level asynchronous programming model which allows for easily writing asynchronous code without having to juggle callbacks or conform to a library which supports asynchronicity.
+
+The core it all are the ``Task`` and ``Task<T>`` objects, which model arbitrary asynchronous operations.  These are interacted with via the ``async`` and ``await`` keywords (``Async`` and ``Await`` in VB) to provide a natural-feeling way to write async code.  The result is the ability to write powerful code which cleanly expresses intent.  There are other ways to approach this than ``async`` and ``await``, outlined in the `Task-based Asynchronous Pattern (TAP) <https://msdn.microsoft.com/en-us/library/hh873175(v=vs.110).aspx>`_, which presents Task-based asynchronocity with options other than the language-level abstractions.  This document will focus on the language-level constructs from this point forward.
+
+For example, say you want to download some data from a web service when a button is pressed, but don't want to hang up the UI thread.  It can be accomplished simply like this:
+
+.. code-block:: c#
+
+	private readonly HttpClient _httpClient = new HttpClient();
+
+	...
+
+	button.Clicked += async (o, e) =>
+	{
+	    var stringData = await _httpClient.DownloadStringAsync(URL);
+	    DoStuff(stringData);
+	};
+
+And that's it!  The code expresses the intent (downloading some data asynchronously) without getting bogged down in interacting with Task objects.
+
+For those who are more theoretically-inclined, this is an implementation of the `Future/Promise concurrency model <https://en.wikipedia.org/wiki/Futures_and_promises>`_.
+
+A few important things to know before continuing:
+
+* Async code uses ``Task<T>`` and ``Task``, which are constructs used to model the work being done in an asynchronous context.  `More on Task and Task<T>`_
+* When the ``await`` keyword is applied, it suspends the calling method and returns control back to its caller until the awaited task is complete.  This is what allows a UI to be responsive and a service to be elastic.
+* ``await`` can only be used inside an async method.
+* Unless an async method has an ``await`` inside its body, it will never yield!
+* ``async void`` should **only** be used on Event Handlers (where it is required).
+
+Example (C#)
+------------
+
+The following example shows how to write basic async code for both a client app and a web service.  The code, in both cases, will count the number of times "Microsoft" appears in the HTML of "microsoft.com".
+
+Client app snippet (Universal Windows App):
+
+.. code-block:: c#
+
+	private readonly HttpClient _httpClient = new HttpClient();
+
+	private async void SeeTheMicrosofts_Click(object sender, RoutedEventArgs e)
+	{
+	    // Capture the task handle here so we can await the background task later.
+	    var getMicrosoftDotComHtmlTask = _httpClient.GetStringAsync("microsoft.com");
```

Comment:
Does that actually work, without the "http://"? Even if it does, I suspect it isn't best practice.